src/main/java/org/folio/circulation/domain/LoanCheckInService.java

```diff
@@ -1,28 +1,53 @@
 package org.folio.circulation.domain;
 
+import static org.folio.circulation.domain.LoanAction.ResolveClaimedReturned;
 import static org.folio.circulation.support.Result.of;
+import static org.folio.circulation.support.Result.succeeded;
 
 import org.folio.circulation.domain.representations.CheckInByBarcodeRequest;
 import org.folio.circulation.support.Result;
-
 import org.joda.time.DateTime;
 
 public class LoanCheckInService {
 
   public Result<Loan> checkIn(Loan loan, DateTime systemDateTime,
-    CheckInByBarcodeRequest request) {
-    return of(() -> loan == null
-      ? null
-      : loan.checkIn(request.getCheckInDate(), systemDateTime,
-        request.getServicePointId()));
+                              CheckInByBarcodeRequest request) {
+
+    if (loan == null) {
+      return of(() -> null);
+    }
+
+    if (loan.getItem().isClaimedReturned()) {
+      return succeeded(loan
+        .resolveClaimedReturned(getClaimedReturnedResolveAction(request),
+          request.getCheckInDate(), systemDateTime, request.getServicePointId()));
+    }
+
+    return succeeded(loan.checkIn(request.getCheckInDate(), systemDateTime,
+      request.getServicePointId()));
   }
 
   public boolean isInHouseUse(Item item, RequestQueue requestQueue,
-    CheckInByBarcodeRequest checkInByBarcodeRequest) {
+                              CheckInByBarcodeRequest checkInByBarcodeRequest) {
 
     return item.isAvailable()
       && requestQueue.getRequests().isEmpty()
       && item.getLocation().homeLocationIsServedBy(checkInByBarcodeRequest
       .getServicePointId());
   }
+
+  private ResolveClaimedReturned getClaimedReturnedResolveAction(
+    CheckInByBarcodeRequest request) {
+
+    switch (request.getClaimedReturnedResolvedBy()) {
+      case FOUND_BY_LIBRARY:
+        return ResolveClaimedReturned.CHECKED_IN_FOUND_BY_LIBRARY;
+      case RETURNED_BY_PATRON:
+        return ResolveClaimedReturned.CHECKED_IN_RETURNED_BY_PATRON;
+      default:
+        throw new IllegalArgumentException(
+          "No loan action to resolve claimed returned: "
+            + request.getClaimedReturnedResolvedBy());
+    }
```

Comment:
What catches this exception?

Reply:
It should be interpreted to a server error, I think. The main purpose of the exception is to notify early if there is a new claimed returned resolution type but it has not mapped to a specific loan action.

Reply:
Please can the calling code for this method be checked that it does interpret this exception as a server error.

Reply:
Unfortunately it is not possible to test this, because we have to add a new value for `ClaimedReturnedResolution` enum. which is not possible to achieve through tests. 

I've changed it to return a result instead.

Reply:
I didn't necessarily mean an automated test, rather an inspection of the code paths where this was used.